jicofo/src/main/kotlin/org/jitsi/jicofo/rest/move/MoveEndpoints.kt

```diff
@@ -0,0 +1,227 @@
+     */
+    @Path("move-fraction")
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    fun moveFraction(
+        /**
+         * Bridge JID, e.g. jvbbrewery@muc.jvb.example.com/jvb1. This is a required parameter, but without a
+         * @DefaultValue jetty returns a 500 and prints a stack trace.
+         */
+        @QueryParam("bridge") bridgeId: String,
+        /** The fraction of endpoints to move. Defaults to 10% */
+        @QueryParam("fraction") @DefaultValue("0.1") fraction: Double
+    ): Result {
+        val bridge = getBridge(bridgeId)
+        val bridgeConferences = bridge.getConferences()
+        val totalEndpoints = bridgeConferences.sumOf { it.second }
+        val numEndpoints = (fraction * totalEndpoints).roundToInt()
+        logger.info("Moving $fraction of endpoints from bridge=$bridge")
+        logger.info("Moving $numEndpoints out of $totalEndpoints")
+        val endpointsToMove = bridgeConferences.select(numEndpoints)
+        return doMove(bridge, endpointsToMove)
+    }
+
+    private fun doMove(bridge: Bridge, endpointsToMove: Map<JitsiMeetConference, Int>): Result {
+        logger.info("Moving endpoints from bridge ${bridge.jid}: $endpointsToMove")
+        var movedEndpoints = 0
+        var conferences = 0
+        endpointsToMove.forEach { (conference, numEps) ->
+            val moved = conference.moveEndpoints(bridge, numEps)
+            movedEndpoints += moved
+            if (moved > 0) conferences++
+        }
+        logger.info("Moved $movedEndpoints endpoints from $conferences conferences.")
+        return Result(movedEndpoints, conferences)
+    }
+
+    private fun getBridge(bridge: String): Bridge {
+        val bridgeJid = try {
+            JidCreate.from(bridge)
+        } catch (e: Exception) {
+            throw BadRequestExceptionWithMessage("Invalid bridge ID")
+        }
+
+        return bridgeSelector.get(bridgeJid) ?: throw NotFoundExceptionWithMessage("Bridge not found")
+    }
+
+    private fun getConference(conferenceId: String): JitsiMeetConference {
+        val conferenceJid = try {
+            JidCreate.entityBareFrom(conferenceId)
+        } catch (e: Exception) {
+            throw BadRequestExceptionWithMessage("Invalid conference ID")
+        }
+        return conferenceStore.getConference(conferenceJid) ?: throw NotFoundExceptionWithMessage("Conference not found")
+    }
+
+    private fun Bridge.getConferences() = conferenceStore.getAllConferences().mapNotNull { conference ->
+        conference.bridges[this]?.participantCount?.let { Pair(conference, it) }
+    }.sortedByDescending { it.second }
+}
+
+data class Result(
+    val movedEndpoints: Int,
+    val conferences: Int
+)
+
+class MoveEndpointsConfig {
+    companion object {
+        val enabled: Boolean by config {
+            "jicofo.rest.move-endpoints.enabled".from(JitsiConfig.newConfig)
+        }
+    }
+}
+
+/**
+ * Select endpoints to move, e.g. with a map m={a: 1, b: 3, c: 3}:
```

Comment:
This could use an explanation of the goal of the algorithm, and why it's a good choice for how to select which participants to move.

As we discussed this might have thundering horde issues, so we should keep an eye on it.